### ISSUE TITLE
Type Foo cannot be assigned to type Foo

### ISSUE THREAD

**Issue by Timmmm** [microsoft/pyright#2333]
With the following code:

```python
class Ob:
    subobjects: list[Ob]

    def split(self) -> list[Ob]:
        result = [self] if len(self.subobjects) > 1 else []
        return result + self.subobjects
```

I get this rather confusing error for the last line:

```
Expression of type "list[Ob]" cannot be assigned to return type "list[Ob]"
  TypeVar "_T@list" is invariant
    Type "Ob" cannot be assigned to type "Ob"
(variable) subobjects: list[Ob]
```

If you hover `result` it tells you the inferred type is `list[Ob]`, however if you explicitly add that

            result: list[Ob] = [self] if len(self.subobjects) > 1 else []

then the error goes away. Seems like there is more to the type than what is printed.

Found via Pylance v2021.9.3.

**Comment by erictraut** [comment]
Yeah, I agree this is confusing.

This stems from the fact that the type of `self` is inferred to be a generic type that is bound to `Ob`. In other words, the type checker internally creates a type variable, similar to this:

```python
__type_of_self_Ob = TypeVar("__type_of_self_Ob", bound="Ob")

class Ob:
    subobjects: list["Ob"]

    def split(self: __type_of_self_Ob) -> list["Ob"]:
        result = [self] if len(self.subobjects) > 1 else []
        return result + self.subobjects
```

Because this TypeVar is generated internally, we don't print it in type messages. Instead, we print the type that it is bound to.

I'll investigate whether there's a way to avoid this false positive and the confusing error message that it's generating.

In the meantime, you can work around it by adding a type annotation to the `result` variable: `result: list["Ob"] = ...`.

**Comment by erictraut** [comment]
I'm struggling to come up with a good solution here. The current implementation is designed to handle cases like this.

```python
class Parent:
    def foo(self):
        return [self]

class Child(Parent): ...

x = Child().foo()
reveal_type(x)  # list[Child]
```

The above code is such an unusual edge case, I've decided to leave the current behavior in place and say that you need to use a type annotation in this case. However, that doesn't address the confusing error message. To address this, I've changed the code that converts the type to a textual representation to include a tilde character ("~") before the class name to indicate that it's a TypeVar that's bound to the class rather than the class itself. This isn't a perfect solution, but it will give the user a clue that something is going on and give them an opportunity to look in the documentation for the meaning of the tilde.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.171 which was just published. It will also be included in the next release of pylance.
